design/authenticators/authn_sut/authn_sut_solution_design.md

```diff
@@ -0,0 +1,433 @@
+| Generate SUT - Invalid code challenge algorithm | Sad | Verify that a SUT cannot be generated with an invalid code challenge algorithm | | |
+| Authenticate with SUT | Happy | Verify that a user can authenticate to Conjur with a SUT | | |
+| Authenticate with SUT - Invalid SUT | Sad | Verify that a user cannot authenticate to Conjur with an invalid SUT | | |
+| Authenticate with SUT - Expired SUT | Sad | Verify that a user cannot authenticate to Conjur with an expired SUT | | |
+| Authenticate with SUT - Used SUT | Sad | Verify that a user cannot authenticate to Conjur with a used SUT | | |
+| Authenticate with SUT - Authenticator not defined in policy | Sad | Verify that a user cannot authenticate to Conjur with an authenticator that is not defined in policy | | |
+| Authenticate with SUT - Invalid authenticator | Sad | Verify that a user cannot authenticate to Conjur with an invalid authenticator | | |
+| Authenticate with SUT - Disabled authenticator | Sad | Verify that a user cannot authenticate to Conjur with a disabled authenticator | | |
+| Authenticate with SUT - Authenticator not enabled for user | Sad | Verify that a user cannot authenticate to Conjur with a disabled user | | |
+| Authenticate with SUT - Incorrect user | Sad | Verify that a user cannot authenticate to Conjur with a SUT for a different user | | |
+| Authenticate with SUT - Missing code verifier | Sad | Verify that a user cannot authenticate to Conjur without a code verifier | | |
+| Authenticate with SUT - Invalid code verifier | Sad | Verify that a user cannot authenticate to Conjur with an invalid code verifier | | |
+| Generate & Authenticate - Multiple SUTs | Sad | Generate two SUTs for a user. Verify that the first one cannot be used after the second is created. | | |
+
+### Integration Tests
+
+Integration tests should cover the following scenarios:
+
+| Scenario | Path Type | Test Purpose | Failure Scenarios | Details |
+|----------|-----------|--------------|-------------------|---------|
+| Generate SUT | Happy | Verify that a SUT can be generated for a user | | |
+| Generate SUT - Invalid credentials | Sad | Verify that a SUT cannot be generated for a user with invalid credentials | | |
+| Authenticate with SUT | Happy | Verify that a user can authenticate to Conjur with a SUT | | |
+| Authenticate with SUT - Invalid SUT | Sad | Verify that a user cannot authenticate to Conjur with an invalid SUT | | |
+| Authenticate with SUT - Expired SUT | Sad | Verify that a user cannot authenticate to Conjur with an expired SUT | | |
+| Authenticate with SUT - Invalid code verifier | Sad | Verify that a user cannot authenticate to Conjur with an invalid code verifier | | |
+| Authenticate with SUT - Incorrect user | Sad | Verify that a user cannot authenticate to Conjur with a SUT for a different user | | |
+| Generate & Authenticate - Multiple SUTs | Sad | Generate two SUTs for a user. Verify that the first one cannot be used after the second is created. | | |
```

Comment:
In the spirit of Jason's desire to minimize cucumber tests stated above, do we need all these Sad path tests here as well as in unit tests? 1 happy path test (generate the SUT and authenticate) and 1 sad path test to make sure errors make it to where they should be seem like enough given the set of unit tests.